[feat] Add Matrix-Game 2.0

H1yori233 opened the pull request:
## Added

### DiT

- `fastvideo/models/dits/matrix_game/` - Matrix-Game DiT implementation
  - `model.py` - Non-causal MatrixGameWanModel
  - `causal_model.py` - Causal CausalMatrixGameWanModel
  - `action_module.py` - Mouse/keyboard action conditioning module
  - `utils.py` - utils for action parsing

### Pipelines

- `fastvideo/pipelines/basic/matrixgame/` - MatrixGame pipeline folder
  - `matrixgame_i2v_pipeline.py` - MatrixGamePipeline
  - `matrixgame_causal_dmd_pipeline.py` - MatrixGameCausalDMDPipeline

### Multi-Variant Support

- `base_distilled_model` - Universal model (keyboard_dim=4)
- `gta_distilled_model` - GTA driving model (keyboard_dim=2)
- `templerun_distilled_model` - Temple Run model (keyboard_dim=7, no mouse)

### Configs

- `fastvideo/configs/models/dits/matrixgame.py` - MatrixGame model arch config
- `fastvideo/configs/pipelines/wan.py` - MatrixGame pipeline configs
- `fastvideo/configs/pipelines/registry.py` - MatrixGame pipeline config mapping
- `fastvideo/configs/sample/base.py` - Added `allow_early_stop` field
- `fastvideo/configs/sample/registry.py` - MatrixGame sampling config registry
- `fastvideo/configs/sample/wan.py` - MatrixGame sampling config

### Pipeline Stages

- `MatrixGameCausalDenoisingStage` - Causal denoising with action conditioning and early stop support
- `MatrixGameImageEncodingStage` - CLIP encoding (uses hidden_states[-2])
- `MatrixGameImageVAEEncodingStage` - VAE encoding, generates cond_concat

### Examples

- `examples/inference/basic/basic_matrixgame.py` - inference with multi-variant support

### Tests

- `fastvideo/tests/transformers/test_matrixgame.py` - MatrixGame model unit tests, **skip**
- `fastvideo/tests/ssim/test_matrixgame_similarity.py` - MatrixGame SSIM test, create a new script to avoid massive conditional branches in a shared SSIM test.
---

## Changed

MatrixGame model and pipeline registration

### ~~CLIPVisionModel~~

~~- Added `output_hidden_states` param to return all layer hidden states~~
~~- Matrix-Game uses second-to-last layer features~~

### Pipeline Stages

- `input_validation.py` - Added mouse_cond/keyboard_cond/grid_sizes validation
- `conditioning.py` - Skip prompt embedding for MatrixGame (without text module)
- `latent_preparation.py` - Support cond_concat concatenation
- `decoding.py` - VAE denormalization for MatrixGame

### Component Loader

- Support loading MatrixGame model weights
- Handle action_config injection

### ForwardBatch

- New fields: `mouse_cond`, `keyboard_cond`, `grid_sizes`

---

## Notes

- The `CausalMatrixGameWanModel` is adapted from [causal_wanvideo.py](https://github.com/hao-ai-lab/FastVideo/blob/main/fastvideo/models/dits/causal_wanvideo.py), and therefore inherits the same architectural constraints:
  - Don't support `DistributedAttention`
  - Only `FLASH_ATTN` and `TORCH_SDPA`
  - Only support single GPU inference
- ~~Matrix-Game uses the **second-to-last hidden state** (`hidden_states[-2]`) instead of the last hidden state for image encoding. The original FastVideo `CLIPVisionModel.forward()` only returned `last_hidden_state` without intermediate layer outputs.~~
- Matrix-Game runs without text prompts (empty `prompt_embeds`), which caused validation failures in the original `ConditioningStage`, so I let it skip prompt embedding and disable CFG.
- `ActionModule` is adapted from the official Matrix-Game repository [`action_module.py`](https://github.com/SkyworkAI/Matrix-Game/blob/main/Matrix-Game-2/wan/modules/action_module.py) with minimal changes for FastVideo integration.
- Upload Matrix-Game-2.0 Diffusers-format to FastVideo HF account.
	- **remove on-the-fly conversion in draft PR**
	- [Matrix-Game-2.0-Base-Diffusers](https://huggingface.co/FastVideo/Matrix-Game-2.0-Base-Diffusers)
	- [Matrix-Game-2.0-GTA-Diffusers](https://huggingface.co/FastVideo/Matrix-Game-2.0-GTA-Diffusers)
	- [Matrix-Game-2.0-TempleRun-Diffusers](https://huggingface.co/FastVideo/Matrix-Game-2.0-TempleRun-Diffusers)
- **No streaming generation** at this time (will leave for another RFC)

H1yori233 committed:
copy-paste action module and wan model, add matrixgame2 initial port

H1yori233 committed:
align pipeline

H1yori233 committed:
mass code changes, hacky pipeline to get matrixgame running...will clean up

H1yori233 committed:
fix temporary hacks and restore original implementation

H1yori233 committed:
ugly conversion and inference scripts

H1yori233 committed:
Merge remote-tracking branch 'origin/main' into feat/kaiqin/add-matrixgame2

H1yori233 committed:
add matrix game causal denoising stages

H1yori233 committed:
fix bug and new logging

H1yori233 committed:
made it runnable

H1yori233 committed:
fix crucial bug:
1. wrong scheduler
2. ff_output calculation
3. num_frame_per_block pass

now it seems normal on first 48 frames, but then still crash to noise

H1yori233 committed:
fix 49frame crash bug!!!

H1yori233 committed:
remove logging

H1yori233 committed:
inference scripts matches official's

H1yori233 committed:
fix param_names_mapping

H1yori233 committed:
fix local_attn_size and num_frame_per_block setting bug by hardcoding them in config

H1yori233 committed:
Merge remote-tracking branch 'origin/main' into feat/kaiqin/add-matrixgame2-debug

H1yori233 committed:
support keyboard input visualization

H1yori233 committed:
draft fix: uses matrix-game version rope, uses block-wise mask

H1yori233 committed:
some fix in random to align with matrix game, but will revert these changes finally

H1yori233 committed:
compare script

H1yori233 committed:
add new log and try to align with official implementation as much as possible

H1yori233 committed:
add custom cross-attention to support cache

H1yori233 committed:
fix clip bug and now it fine with long video

H1yori233 committed:
Merge remote-tracking branch 'origin/main' into feat/kaiqin/add-matrixgame2-debug

H1yori233 committed:
use local attention in action module

H1yori233 committed:
remove posemb_layer

H1yori233 committed:
support streaming generation

H1yori233 committed:
handle model download and conversion automatically.

H1yori233 committed:
code cleanup

H1yori233 committed:
model variant support

H1yori233 committed:
Merge remote-tracking branch 'origin/main' into feat/kaiqin/add-matrixgame2-debug

H1yori233 committed:
move pipeline to matrixgame folder

H1yori233 committed:
Comment out drawing functions to eliminate cv2/libGL dependency in utils.py

H1yori233 committed:
Merge remote-tracking branch 'origin/main' into feat/kaiqin/add-matrixgame2-debug

H1yori233 committed:
remove streaming generation (leave for another RFC)

H1yori233 committed:
remove on-the-fly conversion

H1yori233 committed:
some fix

H1yori233 committed:
fix pre-commit formatting and lint errors

H1yori233 committed:
update reference latent and skip test

H1yori233 committed:
trigger ci

H1yori233 committed:
add ssim for matrixgame & fix pre-commit issue

H1yori233 committed:
move models to FastVideo hf account

H1yori233 committed:
place Diffusers as suffix

SolitaryThinker commented on fastvideo/models/encoders/clip.py:
```diff
@@ -581,11 +581,30 @@ def forward(
+            return BaseEncoderOutput(
+                last_hidden_state=last_hidden_state,
+                hidden_states=tuple(hidden_states_tuple)
+            )
```
You should take a look at this suggestion.

SolitaryThinker commented on fastvideo/models/loader/component_loader.py:
```diff
@@ -323,13 +323,27 @@ def load_model(self,
+        # MatrixGame's CLIP config has fields incompatible with FastVideo's CLIPVisionArchConfig
+        # Use HFCLIPVisionModel directly for MatrixGame to avoid config parsing issues
+        pipeline_class_name = type(fastvideo_args.pipeline_config).__name__
+        if "MatrixGame" in pipeline_class_name:
```
One solution to avoid hardcoding is to directly modify your HF repo's configs and model_index.json to use `HFCLIPVisionModel` instead since our loader uses those to detect the class to instantiate

here https://huggingface.co/FastVideo/Matrix-Game-2.0-Base-Diffusers/blob/main/model_index.json#L18

and here https://huggingface.co/FastVideo/Matrix-Game-2.0-Base-Diffusers/blob/main/image_encoder/config.json#L3

SolitaryThinker commented on fastvideo/pipelines/stages/causal_denoising.py:
Can you just revert the entire class for `CausalDMDDenosingStage` since now you created a new denoising stage for matrix game?

H1yori233 committed:
1. remove hardcode in basic_matrixgame.py
2. add matrix_denoising stage and revert changes in casual_denoising.py
3.  use WAN2_1ControlCLIPVisionConfig in MatrixGameI2V480PConfig to get second-to-last hidden states and revert changes in clip.py
4. revert changes in component_loader.py

H1yori233 committed:
remove manual seed setting in InputValidationStage

H1yori233 committed:
use FastVideo's CLIP instread of HF CLIP will introduces difference, therefore update the ssim test

H1yori233 committed:
trigger ci
